### PR TITLE
Adds a time boxed sampling for backend listeners

### PR DESCRIPTION
**Pull request by max3163**

### CODE REVIEW

File: src/components/org/apache/jmeter/visualizers/backend/WindowMode.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ */
+
+package org.apache.jmeter.visualizers.backend;
+
+import org.apache.jmeter.util.JMeterUtils;
+
+/**
+ * Metrics are sent into boxes which can be {@link #FIXED a fixed-size sliding window} or {@link #TIMED time boxed}.
+ * @author Logan Mzz
```
Comment (FSchumacher) [inline src/components/org/apache/jmeter/visualizers/backend/WindowMode.java]: We don't use the author doc tag.

File: src/components/org/apache/jmeter/visualizers/backend/SamplerMetric.java [hunk 2]
```diff
@@ -88,8 +110,18 @@ private void addHits(SampleResult res) {
...
+    		case FIXED:
+                // We don't clear responsesStats nor usersStats as it will slide as per my understanding of 
+                // http://commons.apache.org/proper/commons-math/userguide/stat.html
+    			break;
+    		case TIMED:
+    			for (DescriptiveStatistics stat : windowedStats) {
+    				stat.clear();
+    			}
+    			break;
+			default: throw new UnsupportedOperationException(windowMode.name());
```
Comment (FSchumacher) [inline src/components/org/apache/jmeter/visualizers/backend/SamplerMetric.java]: Seems like a formatting error

File: src/components/org/apache/jmeter/visualizers/backend/SamplerMetric.java [hunk 3]
```diff
@@ -88,8 +110,18 @@ private void addHits(SampleResult res) {
...
     public synchronized void resetForTimeInterval() {
-        // We don't clear responsesStats nor usersStats as it will slide as per my understanding of 
-        // http://commons.apache.org/proper/commons-math/userguide/stat.html
+    	switch (windowMode) {
+    		case FIXED:
+                // We don't clear responsesStats nor usersStats as it will slide as per my understanding of 
+                // http://commons.apache.org/proper/commons-math/userguide/stat.html
+    			break;
+    		case TIMED:
+    			for (DescriptiveStatistics stat : windowedStats) {
```
Comment (FSchumacher) [inline src/components/org/apache/jmeter/visualizers/backend/SamplerMetric.java]: JMeter now uses java 8, so we could use the newer syntax :)

File: src/components/org/apache/jmeter/visualizers/backend/SamplerMetric.java [hunk 4]
```diff
@@ -28,24 +34,39 @@
...
+    	List<DescriptiveStatistics> stats = new ArrayList<>();
+    	stats.add(pctResponseStats);
+    	if (windowMode == WindowMode.FIXED) {
+    		windowedStats = asList(okResponsesStats, koResponsesStats, allResponsesStats);
+    		stats.addAll(windowedStats);
+    	}
+
+    	int slidingWindowSize = JMeterUtils.getPropDefault("backend_metrics_window", 100);
+    	// Limit to sliding window of SLIDING_WINDOW_SIZE values
+		for (DescriptiveStatistics stat : stats) {
```
Comment (FSchumacher) [inline src/components/org/apache/jmeter/visualizers/backend/SamplerMetric.java]: Formatting issue

File: bin/jmeter.properties [hunk 5]
```diff
@@ -917,6 +917,8 @@ summariser.name=summary
 # BackendListener - configuration
 #---------------------------------------------------------------------------
 #
+# Backend metrics window mode (fixed=fixed-size window, timed=time boxed) 
+#backend_metrics_window_mode=fixed
```
Comment (FSchumacher) [inline bin/jmeter.properties]: Documentation is missing for this new feature. It should be mentioned in the properties section and the users manual.

**Review by FSchumacher — CHANGES_REQUESTED** [review]
Thanks for the contribution.
Overall it looks good. Maybe you could also describe the use case for this feature compared to the fixed-size one.

**Comment by pmouawad** [comment]
Hello @max3163,
Thank you for your contribution.
Could you (as requested by Felix) give more details on the use case for this PR ?
Thank you
Regards

**Comment by max3163** [comment]
When you use the fixed metric window mode, all samples metrics are saved in a sliding array with a fix size, you send each second to your backend a trend of this values
When you use the time boxed mode, all samples metrics are reset in order to send actual values for the last elapse time.
I you use it with a Influxdb backend (I'm thinking to send a PR to share it too) with a pooling of 5 seconds.

**Comment by pmouawad** [comment]
Hello Team,
PR looks fine to me.
Do you agree with enhancement ?
Thank you

**Comment by Wyatts** [comment]
This sounds good if it's what I think it is.  Definitely would appreciate a clear description of what it precisely does, though.

**Comment by pmouawad** [comment]
@max3163 , are you using GraphiteBackendListener or something else (I don't fully understand your comment on "I'm thinking to send a PR to share it too".
If not, why aren't you using it ?
Thanks

**Comment by max3163** [comment]
@pmouawad 
We use ur own BackendListener to send metric on InfluxDB with his HTTP API and follow his own shema design ( see : [Influxdb data design](https://docs.influxdata.com/influxdb/v1.1/concepts/schema_and_data_layout/#encouraged-schema-design) ) which is more better than use the Graphite plugin.
If you want, we can share this BackendListener.

**Comment by pmouawad** [comment]
Hello,
Thanks for proposal.
What is the quantity of code ? Does it have any dependency on Influx library or it is standalone using socket ?

Thank you

**Comment by max3163** [comment]
Hi, 

I've just send you a new PR ( #246 )for the Influxdb backend. 
It's about 650 lines of code ( test class included ) and with no dependency on Influxdb library.

**Pull request closed without merging**